File: src/main/kotlin/dev/moru3/compsql/datatype/DataType.java
package dev.moru3.compsql.datatype;

import dev.moru3.compsql.TypeHub;
import dev.moru3.compsql.datatype.types.binary.*;
import dev.moru3.compsql.datatype.types.date.*;
import dev.moru3.compsql.datatype.types.bool.*;
import dev.moru3.compsql.datatype.types.text.*;
import dev.moru3.compsql.datatype.types.numeric.*;
import dev.moru3.compsql.datatype.types.numeric.unsigned.*;
import dev.moru3.compsql.datatype.types.original.*;

public class DataType {
    // binary系
    public static BINARY BINARY = TypeHub.INSTANCE.register(new BINARY(255));
    public static LONGBLOB LONGBLOB = TypeHub.INSTANCE.register(new LONGBLOB(1023));
    public static VARBINARY VARBINARY = TypeHub.INSTANCE.register(new VARBINARY(255));

    // boolean系
    public static BOOLEAN BOOLEAN = TypeHub.INSTANCE.register(new BOOLEAN());

    // numeric系
    public static BIGINT BIGINT = TypeHub.INSTANCE.register(new BIGINT((byte) 19));
    public static INTEGER INTEGER = TypeHub.INSTANCE.register(new INTEGER((byte) 10));
    public static SMALLINT SMALLINT = TypeHub.INSTANCE.register(new SMALLINT((byte) 5));
    public static TINYINT TINYINT = TypeHub.INSTANCE.register(new TINYINT((byte) 3));
    // unsigned
    public static UBIGINT UBIGINT = TypeHub.INSTANCE.register(new UBIGINT((byte) 20));
    public static UINTEGER UINTEGER = TypeHub.INSTANCE.register(new UINTEGER((byte) 10));
    public static USMALLINT USMALLINT = TypeHub.INSTANCE.register(new USMALLINT((byte) 5));
    public static UTINYINT UTINYINT = TypeHub.INSTANCE.register(new UTINYINT((byte) 3));

    // text系
    public static CHAR CHAR = TypeHub.INSTANCE.register(new CHAR(255));
    public static LONGTEXT LONGTEXT = TypeHub.INSTANCE.register(new LONGTEXT(1023));
    public static TEXT TEXT = TypeHub.INSTANCE.register(new TEXT(511));
    public static VARCHAR VARCHAR = TypeHub.INSTANCE.register(new VARCHAR(255));

    // date型
    public static DATE DATE = TypeHub.INSTANCE.register(new DATE(null));
    public static DATETIME DATETIME = TypeHub.INSTANCE.register(new DATETIME(null));
}

File: src/main/kotlin/dev/moru3/compsql/datatype/DataType.kt
package dev.moru3.compsql.datatype
import dev.moru3.compsql.datatype.types.binary.*
import dev.moru3.compsql.datatype.types.bool.*
import dev.moru3.compsql.datatype.types.date.DATE
import dev.moru3.compsql.datatype.types.date.DATETIME
import dev.moru3.compsql.datatype.types.text.*
import dev.moru3.compsql.datatype.types.numeric.*
import dev.moru3.compsql.datatype.types.numeric.unsigned.*
import org.jetbrains.annotations.Contract
import java.sql.*

class Range(val a: Int, val b: Int) {
    override fun toString(): String {
        return "$a, $b"
    }
}

/**
 * 新しいSQLTypeを作成するにはこれを使用してください。
 */
interface DataType<F, T> {

    val typeName: String
    val from: Class<F>
    val type: Class<T>
    val sqlType: Int
    val allowPrimaryKey: Boolean
    val allowNotNull: Boolean
    val allowUnique: Boolean
    val isUnsigned: Boolean
    val allowZeroFill: Boolean
    val allowAutoIncrement: Boolean
    val allowDefault: Boolean
    val defaultProperty: String?
    val priority: Int

    /**
     * PreparedStatementの指定されたindexに対してanyをsetします。
     * anyがDataTypeのtypeと異なる場合はIllegalStateExceptionをthrowします。
     * @param ps setするPreparedStatement
     * @param index setする変数のindex
     * @param any setするオブジェクト。
     */
    @Contract("_, _, null -> !null")
    fun set(ps: PreparedStatement, index: Int, any: Any?) {
        check(from.isInstance(any)) { "The type of \"${if(any!=null) any::class.javaObjectType.simpleName else "null"}\" is different from \"${type::class.javaObjectType.simpleName}\"." }
        ps.setObject(index, any, sqlType)
    }

    /**
     * ResultSetからidに対応した値を変換して返します。
     */
    fun get(resultSet: ResultSet, id: String): F?

    companion object {
        // binary系
        val BINARY = BINARY(255)
        val LONGBLOB = LONGBLOB(1023)
        val VARBINARY = VARBINARY(255)

        // boolean系
        val BOOLEAN = BOOLEAN()

        // numeric系
        val BIGINT = BIGINT(19)
        val INTEGER = INTEGER(10)
        val SMALLINT = SMALLINT(5)
        val TINYINT = TINYINT(3)
        // unsigned
        val UBIGINT = UBIGINT(20)
        val UINTEGER = UINTEGER(10)
        val USMALLINT = USMALLINT(5)
        val UTINYINT = UTINYINT(3)

        // text系
        val CHAR = CHAR(255)
        val LONGTEXT = LONGTEXT(1023)
        val TEXT = TEXT(511)
        val VARCHAR = VARCHAR(255)

        // date型
        val DATE = DATE(null)
        val DATETIME = DATETIME(null)
    }
}